cachi2/core/safepath.py

```diff
@@ -22,38 +16,65 @@ def __init__(self, reason: str) -> None:
         super().__init__(reason, solution=None, docs=None)
 
 
-class SafePath(PathType):
-    """A Path subclass with a safer joinpath.
+class SafePath(PathLike[str]):
+    """A safer way to handle subpaths.
+
+    Get a subpath, guaranteeing that it really is a subpath:
 
     >>> safe_path = SafePath("/some/directory")
+    >>> safe_path.safe_join("..")                   # ERROR NotSubpath
+    >>> safe_path.safe_join("/abspath")             # ERROR NotSubpath
+    >>> safe_path.safe_join("symlink-to-parent")    # ERROR NotSubpath
+
+    Access the underlying Path object:
 
-    Subpaths obtained via `safe_path / path` or `safe_path.joinpath(*path_components)`
-    must be subpaths of the safe_path. Otherwise, the join operation raises the NotSubpath
-    error.
+    >>> safe_path = SafePath("/some/directory")
+    >>> safe_path.safe_join("vendor", "modules.txt").path.read_text()
     """
 
-    def __new__(cls: type[SafePathT], abspath: StrPath) -> SafePathT:
+    def __init__(self, path: StrPath) -> None:
         """Create a SafePath. The argument must be an absolute path."""
-        path = super().__new__(cls, abspath)
-        if not path.is_absolute():
-            raise ValueError(f"safe path must be absolute but isn't: {abspath}")
-        return path
+        self._path = Path(path)
+        if not self._path.is_absolute():
+            raise ValueError(f"path must be absolute: {path}")
+
+    @property
+    def path(self) -> Path:
+        """Get the underlying Path object."""
+        return self._path
+
+    def __fspath__(self) -> str:
+        return self.path.__fspath__()
 
-    def joinpath(self: SafePathT, *other: StrPath) -> SafePathT:
+    def __str__(self) -> str:
+        return str(self.path)
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}({str(self.path)!r})"
+
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, type(self)) and self.path == other.path
+
+    def __hash__(self) -> int:
+        return hash(self.path)
+
+    def safe_join(self: SafePathT, *other: StrPath) -> SafePathT:
```

Comment:
If we do go for the explict approach, I'm wondering if `Safepath.safe_join` should return a Safepath at all. A dev can mistakenly do:

```python
path = SafePath("/root/path")
another_path = path.safe_join("submodule")
another_path.join("..") # would fail, since another_path's root is '/root/path/submodule'
``` 

We could keep track of the root path, but that is a potentially confusing behavior. If we make the `safe_join` method return a `Path`, in order to set a new root path, we would need to explicity instantiate a new SafePath.

We also would save the `.path` part from results of `safe_joins`. The downside is that one could expect that a `safe_join` would return a SafePath.